crates/matrix-sdk/src/guaranteed_completion.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2022 Famedly GmbH
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Optional methods guaranteeing a synchronized event store upon return.
+//!
+//! # How it works
+//! For each method with the suffix `_completed`, an event handler is
+//! registered before the sister function without the suffix is called.
+//! Upon return of the latter, a channel to the event handler receives
+//! events related to the action requested from the Matrix Server and
+//! the function returns only after the corresponding event has been
+//! registered.
+
+use ruma::{
+    api::client::membership::{join_room_by_id, join_room_by_id_or_alias},
+    events::{
+        room::member::{MembershipState, RoomMemberEventContent},
+        StateEvent, SyncStateEvent,
+    },
+    OwnedServerName, RoomId, RoomOrAliasId,
+};
+use tokio::sync::mpsc;
+use tracing::debug;
+
+use crate::{
+    error::HttpResult,
+    room::{Common, Room},
+    Client, Result,
+};
+
+impl Client {
+    /// Join a room by `RoomId` and return only after the corresponding
+    /// event has been synced. Requires a running Sync.
+    ///
+    /// Returns a `join_room_by_id::Response` consisting of the
+    /// joined rooms `RoomId`.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_id` - The `RoomId` of the room to be joined.
+    pub async fn join_room_by_id_completed(
+        &self,
+        room_id: &RoomId,
+    ) -> HttpResult<join_room_by_id::v3::Response> {
```

Comment:
Given the introduced functionality relies on event handlers that fire after the state was updated, we could probably also change the API of the methods to return room structs instead of http response. So e.g. here `join_room_by_id_completed` could directly return a `JoinedRoom`, which probably would make sense from the API perspective, since the reason to wait for the call to complete is probably to get the joined room anyway.

Reply:
I could implement this in two different ways.

Either I send the `Room` instance passed to the event handler via the channel together with the full event as a tuple and create a `JoinedRoom` with this room instance, or I use the `room_id` in the API response to call `self.get_joined_room`. 

Any preferences, @jplatte ?

Reply:
I don't have a preference, no. Maybe @poljar does.

Reply:
I've settled with the former approach.